File: algorithms/377.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
private:
    unordered_map<int, int> m;
    int helper(const vector<int> &nums, int target) {
        if (target <= 0)
            return target == 0 ? 1 : 0;
        if (m.find(target) != m.end())
            return m[target];
        int result = 0;
        for (int i = 0; i < nums.size(); i++)
            result += helper(nums, target-nums[i]);
        m[target] = result;
        return result;
    }
public:
    int combinationSum4(vector<int> &nums, int target) {
        m.clear();
        return helper(nums, target);
    }
};

int main() {
    return 0;
}

File: algorithms/rust/377.rs
impl Solution {
    #[allow(dead_code)]
    pub fn combination_sum4(nums: Vec<i32>, target: i32) -> i32 {
        let target = target as usize;
        let mut dp = vec![0; target + 1];
        dp[0] = 1;

        for i in 1..=target {
            for &num in nums.iter() {
                let num = num as usize;
                if i >= num {
                    dp[i] += dp[i - num];
                }
            }
        }

        dp[target]
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
